src/character.cpp

```diff
@@ -8884,19 +8888,103 @@ bool Character::in_sleep_state() const
 
 bool Character::has_item_with_flag( const flag_id &flag, bool need_charges ) const
 {
-    return has_item_with( [&flag, &need_charges, this]( const item & it ) {
-        if( it.is_tool() && need_charges ) {
-            return it.has_flag( flag ) && ( it.type->tool->max_charges == 0 ||
-                                            it.ammo_remaining( this ) > 0 );
+    for( const item *it : all_items_with( flag ) ) {
+        if( !need_charges || !it->is_tool() || it->type->tool->max_charges == 0 ||
+            it->ammo_remaining( this ) > 0 ) {
+            return true;
+        }
+    }
+    return false;
+}
+
+std::vector<item *> &Character::all_items_with( const flag_id &flag ) const
+{
+    std::string key = string_format( "HAS FLAG %s", flag.c_str() );
+
+    auto iter = inv_search_caches.find( key );
+    if( iter != inv_search_caches.end() ) {
+        // If the cache already exists, use it.
+        return iter->second.items;
+    }
+    // Otherwise, add a new cache and populate with all appropriate items in the inventory. Empty sets are still created.
+    inv_search_caches[key].flag = flag;
+    visit_items( [this, &key, &flag]( item * it, item * ) {
+        if( it->has_flag( flag ) ) {
+            inv_search_caches[key].items.push_back( it );
+        }
+        return VisitResponse::NEXT;
+    } );
+
+    return inv_search_caches[key].items;
```

Comment:
Deduplicate code, see suggestions below too

```suggestion
    return all_items_with( "HAS FLAG " + flag.str(), flag, nullptr );
```